### DISCUSSION TITLE
How do I execute a script at installation time to configure environment variables？

### DISCUSSION THREAD

**Discussion by piperhex** [pypa/setuptools#3164] [Q&A] [answered]
How do I execute a script at installation time to configure environment variables？（when i use `pip install mypackage`）

**Comment by piperhex** [comment]
For example, my program needs to execute some scripts to initialize at install time .

**Comment by abravalheri** [comment, marked as answer]
The "executing arbitrary scripts at install time" use case is not supported by the combination of `setuptools`+`pip`.

You might have some luck if you package your project in a OS-specific way (e.g. deb, rpm, nix, etc...).

Otherwise you can try to:
- create a command to be explicitly run by the users (e.g. `mypackage ensureenv`),
  ([this approach is adopted by pipx](https://pypa.github.io/pipx/))
- ask the users to include a custom command in their `.bashrc/.zshrc/...`
  e.g. `source "$(mypackage init --bashrc)"` or `eval "$(mypackage init --bashrc)"`,
  ([this approach is adopted by pyenv](https://github.com/pyenv/pyenv#basic-github-checkout))
- ask your users to use a bash installation script
  ([pyenv](https://github.com/pyenv/pyenv-installer) also offer this approach as an alternative)

**Reply by abravalheri** [reply]
If the stuff you want to do is not too complex, you can also try to execute the script automatically the first time your program runs...

(The following times you can just check if the stuff you wanted to do is already done and skip this step).

**Reply by piperhex** [reply]
Thank you very much for your reply. These suggestions helped me a lot.
